### PR TITLE
Change method names for javadoc and sources jar creation 

### PR DESCRIPTION
**Pull request by ljacomet**
The new name reflects better what happens exactly when the option is
activated on the java extension.
The publication part is only applicable when a publishing plugin is also
configured.

Also includes:
* packaging of resources in sources jar
* wiring with assemble task
* documentation of these methods for feature variants

### CODE REVIEW

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/internal/JvmPluginsHelper.java [hunk 1]
```diff
@@ -224,12 +225,15 @@ public static void configureDocumentationVariantWithArtifact(String variantName,
...
 
         if (!tasks.getNames().contains(jarTaskName)) {
-            tasks.register(jarTaskName, Jar.class, jar -> {
+            TaskProvider<Jar> jarTask = tasks.register(jarTaskName, Jar.class, jar -> {
                 jar.setDescription("Assembles a jar archive containing the " + (featureName == null ? "main " + docsType + "." : (docsType + " of the '" + featureName + "' feature.")));
                 jar.setGroup(BasePlugin.BUILD_GROUP);
                 jar.from(artifactSource);
                 jar.getArchiveClassifier().set(TextUtil.camelToKebabCase(featureName == null ? docsType : (featureName + "-" + docsType)));
             });
+            if (tasks.getNames().contains(LifecycleBasePlugin.ASSEMBLE_TASK_NAME)) {
```
Comment (ljacomet) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/internal/JvmPluginsHelper.java]: I have been really conservative there. If you believe we can assume `assemble` is always there, the `if` can be removed.

File: subprojects/docs/src/docs/dsl/org.gradle.api.plugins.JavaPluginExtension.xml [hunk 2]
```diff
@@ -46,10 +46,13 @@
...
             <tr>
-                <td>publishJavadoc</td>
+                <td>withJavadocJar</td>
             </tr>
             <tr>
-                <td>publishSources</td>
+                <td>withSourcesJar</td>
+            </tr>
+            <tr>
+                <td>registerFeature</td>
```
Comment (jjohannes) [inline subprojects/docs/src/docs/dsl/org.gradle.api.plugins.JavaPluginExtension.xml]: Good catch!

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPluginExtension.java [hunk 3]
```diff
@@ -75,7 +75,12 @@
     void disableAutoTargetJvm();
 
     /**
-     * Automatically package Javadoc during publishing and publish a variant 'javadocElements' with an artifact '-javadoc.jar'.
+     * Adds a task {@code javadocJar} that will package the output of the {@code javadoc} task in a JAR with classifier {@code javadoc}.
+     * <P>
+     * The produced artifact is registered as a documentation variant on the {@code java} component and added as a dependency on the {@code assemble} task.
+     * This means that if the {@code maven-publish} or {@code ivy-publish} is also applied, the javadoc JAR will be published.
```
Comment (jjohannes) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPluginExtension.java]: :o: remove "the"

or change it to

the maven-publish or ivy-publish **plugin**

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPluginExtension.java [hunk 4]
```diff
@@ -84,10 +89,15 @@
...
     @Incubating
-    void publishJavadoc();
+    void withJavadocJar();
 
     /**
-     * Automatically package source code during publishing and publish a variant 'sourceElements' with an artifact '-sources.jar'.
+     * Adds a task {@code sourcesJar} that will package the Java sources of the main {@link org.gradle.api.tasks.SourceSet SourceSet} in a JAR with classifier {@code sources}.
+     * <P>
+     * The produced artifact is registered as a documentation variant on the {@code java} component and added as a dependency on the {@code assemble} task.
+     * This means that if the {@code maven-publish} or {@code ivy-publish} is also applied, the sources JAR will be published.
```
Comment (jjohannes) [inline subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPluginExtension.java]: :o: same as above

File: subprojects/plugins/src/main/java/org/gradle/api/plugins/internal/JvmPluginsHelper.java (hunk 1 above)
Comment (jjohannes) [reply subprojects/plugins/src/main/java/org/gradle/api/plugins/internal/JvmPluginsHelper.java]: I would still keep that here

**Pull request merged**